### DISCUSSION TITLE
undefined reference to `SPI1'

### DISCUSSION THREAD

**Discussion by TheLardos** [earlephilhower/arduino-pico#2320] [Q&A]
Hi everyone,

I'm trying to get my second SPI interface to work. I include the 

`#include <SPI.h>`

header and in the setup I use:

```
  SPI1.setRX(SD_MISO_PIN);
  SPI1.setTX(SD_MOSI_PIN);
  SPI1.setSCK(SD_SCK_PIN);
  SPI1.begin();
  // more code...
  SPI1.end();
  
  SPI.setSCK(CLOCKPIN);
  SPI.setTX(DATAPIN);
  SPI.begin();
  // more code...
  SPI.end();
```

and I run into the error:

`undefined reference to `SPI1'`

I'm probably missing something quite obvious but I couldn't figure out what it is...
I'm glad for any hel I can get.

Cheers,
Luca

**Comment by earlephilhower** [comment]
The code compiles fine with the core and the RPi Pico/W boards (replacing `#define` constants with actual #s, of course).  There are 2 possibilities:
1. The board you've selected has, for some reason, not defined `SPI1` (missing pins, not labeled on their PCB, etc.)
2. You're not actually using this core, and are using the incompatible Arduino.cc one.

For #1 you can manually instantiate a `SPI1` device, look at how it's done in the code here.
https://github.com/earlephilhower/arduino-pico/blob/28b0edac217c949af3bdfd71174464edfad92b7f/libraries/SPI/src/SPI.cpp#L468-L474

For #2, well, re-select the `Raspberry Pi Pico/RP2040` package in boards manager and work from there. :)

**Comment by TheLardos** [comment]
Thanks for your quick reply!
I'm using the "Pimoroni Tiny2040" Board which at least on the pinout diagram should have two SPI interfaces. 
But you're right, once I switch to the "Generic RP2040" Board, it compiles no problem!

I will check out your solution #1 once I have some time.

Thanks again for your fast help!

**Reply by earlephilhower** [reply]
It's very possible the board definition could be updated to add whatever SPI1 pinout is silkscreened/pinned out.  You can do a PR to set the SPI1 options in the [variant file](https://github.com/earlephilhower/arduino-pico/blob/28b0edac217c949af3bdfd71174464edfad92b7f/variants/pimoroni_tiny2040/pins_arduino.h#L19), if you have the time.  Then the next user will get the right config. automatically.

**Comment by Andy2No** [comment]
The pinout diagram for the Pimoroni Tiny2040 only shows that SPI0 can be mapped to one of two sets of pins, either GP0-3 or GP4-7:

![tiny-2040-pinout-diagra](https://github.com/user-attachments/assets/dab0ed11-b0fc-4274-bc5b-4f061d7bb91d)

Looking at the one for the standard Pi Pico, SPI1 could be mapped to the analog input pins, GP26-29 (the Pico doesn't have GP29 on a header pin):

![pin ref 4883-06](https://github.com/user-attachments/assets/a679a10f-bab5-47fb-8e23-17eab4d7f541)

I don't know if that breaks any rules for pin assignments in Arduino, but it looks like that would be the only option.

**Comment by Andy2No** [comment]
I now see that #define PIN_SPI0_SS  should have been 5 (GP5).  I failed to understand that CSn, on the pinout, was meant to represent SS.

I freely admit I don't know enough about the rules to make these choices and the limited number of pins on the Tiny2040 makes it harder.  I think my reasoning was that Rx and Tx are on GP0-1, so it was better to put the default pins for SPI0 on 4-7.  Since everything has to be somewhere, I also put the default pins for I2C0 on 4-5, like on the Pi Pico since there isn't a choice that doesn't overlap something else, as far as I can see.  Maybe there's a better way to arrange it all though.

**Comment by Andy2No** [comment]
@TheLardos Here's a proposed new variant file for the Pimoroni Tiny2040.  I've zipped it so github can allow me to attach it.

[Tiny2040fix_pins_arduino.zip](https://github.com/user-attachments/files/16532475/Tiny2040fix_pins_arduino.zip)


This defaults to SPI0 on GP4-7, as before, but puts the SS pin on GP5.  It also allows SPI1, and defaults to using the four analog input pins - which is the only place you'd be able to set it, but it doesn't stop your code setting those pins explicitly, so long as it assigns each function to the correct pin.

Here's your test code, with some #defines so that it can be compiled, using this variants file;

```
/* TheLardos
Aug 6, 2024

I'm trying to get my second SPI interface to work. I include the

*/

//<A2N>
// SPI1 only available on analog inputs
//define PIN_SPI1_MISO  (28u)
//deine PIN_SPI1_MOSI  (27u)
//define PIN_SPI1_SCK   (26u)
//define PIN_SPI1_SS    (29u)

#define SD_MISO_PIN (28u)
#define SD_MOSI_PIN (27u)
#define SD_SCK_PIN (26u)

//  SPI.setSCK(CLOCKPIN);
#define CLOCKPIN (6u)
//  SPI.setTX(DATAPIN);
#define DATAPIN (7u)

//</A2N>

#include <SPI.h>

void setup() {

  SPI1.setRX(SD_MISO_PIN);
  SPI1.setTX(SD_MOSI_PIN);
  SPI1.setSCK(SD_SCK_PIN);
  SPI1.begin();
  // more code...
  SPI1.end();
  
  SPI.setSCK(CLOCKPIN);
  SPI.setTX(DATAPIN);
  SPI.begin();
  // more code...
  SPI.end();

}

void loop() {
  // put your main code here, to run repeatedly:

}
```

Place the unzipped variants file in your equivalent of;

C:\Users\andy\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\3.4.1\variants\pimoroni_tiny2040

This is from my Windows 7 machine, which as an older copy of the arduino-pico core installed, so you'll have a different version number instead of 3.4.1.  You can get part of the way to this folder by using File->Preferences in the Arduino IDE then clicking on the link for preferences.txt, near the bottom.  That gets me to C:\Users\andy\AppData\Local\Arduino15.  From there, you open the packages folder, then rp2040... and so on.  You'll see a pins_arduino.h  - rename that, e.g. to pins_arduino.hOLD, then paste in the new, unzipped pins_arduino.h  You should now be able to use it without having to wait for a new release, and you can test it for us.  I've only checked that it now compiles.  I can't currently find my Tiny2040 from last year but I might still get lucky.

Here's the contents of that new (as yet untested) file:

```
#pragma once

// This is a bare board with few predefined pins, so based on generic

// Pin definitions taken from:
//    https://datasheets.raspberrypi.org/pico/pico-datasheet.pdf
//    https://shop.pimoroni.com/products/tiny-2040

// RGB LED on pins 18-20.  Use the Green LED element on pin 19 as LED_BUILTIN
#define PIN_LED        (19u)

// Serial
#define PIN_SERIAL1_TX (0u)
#define PIN_SERIAL1_RX (1u)

#define PIN_SERIAL2_TX (4u)
#define PIN_SERIAL2_RX (5u)

// SPI0 can only be on GPI0 0-3 or 4-7
#define PIN_SPI0_MISO  (4u)
#define PIN_SPI0_MOSI  (7u)
#define PIN_SPI0_SCK   (6u)
#define PIN_SPI0_SS    (5u)

// SPI1 only available on analog inputs
#define PIN_SPI1_MISO  (28u)
#define PIN_SPI1_MOSI  (27u)
#define PIN_SPI1_SCK   (26u)
#define PIN_SPI1_SS    (29u)

// Wire = WIRE0 on 0-1 or 4-5, WIRE1 on 2-3 or 6-7
#define PIN_WIRE0_SDA  (4u)
#define PIN_WIRE0_SCL  (5u)

#define PIN_WIRE1_SDA  (6u)
#define PIN_WIRE1_SCL  (7u)

#define SERIAL_HOWMANY (3u)
#define SPI_HOWMANY    (2u)
#define WIRE_HOWMANY   (2u)

#include "../generic/common.h"
```
